base/loading.jl

```diff
@@ -511,26 +592,55 @@ module_uuid(m::Module) = ccall(:jl_module_uuid, UInt64, (Any,), m)
 
 isvalid_cache_header(f::IOStream) = 0 != ccall(:jl_read_verify_header, Cint, (Ptr{Void},), f.ios)
 
-function cache_dependencies(f::IO)
-    modules = Tuple{Symbol,UInt64}[]
-    files = Tuple{String,Float64}[]
+function parse_cache_header(f::IO)
+    modules = Dict{Symbol,UInt64}()
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
-        push!(modules,
-              (Symbol(read(f, n)), # module symbol
-               ntoh(read(f, UInt64)))) # module UUID (timestamp)
+        sym = Symbol(read(f, n)) # module symbol
+        uuid = ntoh(read(f, UInt64)) # module UUID (mostly just a timestamp)
+        modules[sym] = uuid
     end
-    read(f, Int64) # total bytes for file dependencies
+    totbytes = ntoh(read(f, Int64)) # total bytes for file dependencies
+    # read the list of files
+    files = Tuple{String,Float64}[]
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
+        totbytes -= 4 + n + 8
+        if n < 0 # probably means this wasn't a valid file to be read by Base.parse_cache_header
+            error("EOF while reading cache header")
+        end
         push!(files, (String(read(f, n)), ntoh(read(f, Float64))))
     end
+    @assert totbytes == 4 "header of cache file appears to be corrupt"
```

Comment:
It seems this is not really about assertions, but instead a claim that some condition thought to be impossible is in fact possible, and an important failure mode that needs diagnostics.

The question is, would it have come up if this assertion had not been written at all? People should be free to add as many assertions as possible, the more the merrier, without drawing fire about some assertion needing a better message. Otherwise you're just discouraging writing assertions. But if there is some independent reason to believe this is an important failure mode, then it's reasonable to ask for better diagnostics.